barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.hpp

```diff
@@ -188,19 +190,75 @@ template <typename Flavor> class SumcheckProverRound {
         }
         // For ZK Flavors: The evaluations of the round univariates are masked by the evaluations of Libra univariates
         if constexpr (Flavor::HasZK) {
+            const auto contribution_from_disabled_rows = compute_disabled_contribution(
+                polynomials, relation_parameters, gate_sparators, alpha, round_idx, row_disabling_poly);
             const auto libra_round_univariate = compute_libra_round_univariate(zk_sumcheck_data, round_idx);
             // Batch the univariate contributions from each sub-relation to obtain the round univariate
             const auto round_univariate =
                 batch_over_relations<SumcheckRoundUnivariate>(univariate_accumulators, alpha, gate_sparators);
             // Mask the round univariate
-            return round_univariate + libra_round_univariate;
+            return round_univariate + libra_round_univariate - contribution_from_disabled_rows;
         }
         // Batch the univariate contributions from each sub-relation to obtain the round univariate
         else {
             return batch_over_relations<SumcheckRoundUnivariate>(univariate_accumulators, alpha, gate_sparators);
         }
     }
 
+    /*!
+     * @brief For ZK Flavors: A method disabling the last 4 rows of the ProverPolynomials
+     *
+     * @details See description of RowDisablingPolynomial
+     *
+     */
+    template <typename ProverPolynomialsOrPartiallyEvaluatedMultivariates>
+    SumcheckRoundUnivariate compute_disabled_contribution(
```

Comment:
This function is quite unreadable, I think ideally you would find a way to not have so many if cases and unify functionality which definitely seems possible. Also `compute_univariate` has become quite a complex function and I think the ZK functionality could be isolated in a better way. Let me know if you wanna discuss more about this.

Reply:
oh thanks, made it more uniform